pkg/gadgets/trace/dns/tracer/gc.go

```diff
@@ -0,0 +1,100 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !withoutebpf
+
+package tracer
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/cilium/ebpf"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+// Delay between each garbage collection run.
+const garbageCollectorInterval = 1 * time.Second
+
+// startGarbageCollector runs a background goroutine to delete old query timestamps
+// from the DNS query_map. This ensures that queries that never receive a response
+// are deleted from the map.
+//
+// The garbage collector goroutine terminates when the context is done.
+func startGarbageCollector(ctx context.Context, logger logger.Logger, queryMaxAge time.Duration, queryMap *ebpf.Map) {
+	logger.Debugf("starting garbage collection for DNS tracer with queryMaxAge %s", queryMaxAge)
+	go func() {
+		ticker := time.NewTicker(garbageCollectorInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Debugf("stopping garbage collection for DNS tracer")
+				return
+
+			case <-ticker.C:
+				logger.Debugf("executing DNS query map garbage collection")
+				collectGarbage(logger, queryMaxAge, queryMap)
+			}
+		}
+	}()
+}
+
+func collectGarbage(logger logger.Logger, queryMaxAge time.Duration, queryMap *ebpf.Map) {
+	var (
+		key          dnsQueryKeyT
+		val          dnsQueryTsT
+		keysToDelete []dnsQueryKeyT
+	)
+	cutoffTs := types.Time(time.Now().Add(-1 * queryMaxAge).UnixNano())
+	iter := queryMap.Iterate()
+
+	// If the BPF program is deleting keys from the map during iteration,
+	// we may see duplicate keys or stop without processing some keys (ErrIterationAborted).
+	// Duplicate keys are okay since we handle ErrKeyNotExists on delete,
+	// and ErrIterationAborted is okay because we'll retry on the next garbage collection.
+	for iter.Next(&key, &val) {
+		ts := gadgets.WallTimeFromBootTime(val.Timestamp)
+		if ts < cutoffTs {
+			keysToDelete = append(keysToDelete, key)
+		}
+	}
+
+	if err := iter.Err(); err != nil {
+		if errors.Is(err, ebpf.ErrIterationAborted) {
+			logger.Warnf("received ErrIterationAborted when iterating through DNS query map, possibly due to concurrent deletes. Some entries may be skipped this garbage collection cycle.")
+		} else {
+			logger.Errorf("received err %s when iterating through DNS query map", err)
+		}
+	}
+
+	for _, key := range keysToDelete {
+		logger.Debugf("deleting key with mntNs=%d and DNS ID=%x from query map for DNS tracer", key.MountNsId, key.Id)
```

Comment:
changed it to a warning with the msg "DNS query with ID=%x in mntNs=%d did not receive a response within the DNS timeout (%s)"

Reply:
I don't know. If you're debugging the DNS traffic, I don't think it would be annoying to know that a given request didn't receive a response within X seconds (where X is btw configurable).

Reply:
I agree having a notification about a missing answer is good, but this message indicates something different, it's not that straightforward for the user to understand what's the real meaning of this.

Reply:
I removed this log msg when switching from the map iterator to batch lookup/delete. Will defer adding an event/notification since that's tracked by issue https://github.com/inspektor-gadget/inspektor-gadget/issues/1512